File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string pattern = "";
        Helper(0,0,n,ans,pattern);
        return ans;
    }
    void Helper(int openP,int closeP,int n,vector<string>& ans,string& pattern)
    {
        if(openP==closeP && openP==n){
            ans.push_back(pattern);
            return;
        }
        if(openP < n){
            pattern += '(';
            Helper(openP+1,closeP,n,ans,pattern);
            pattern.pop_back();
        }
        if(closeP < openP){
            pattern += ')';
            Helper(openP,closeP+1,n,ans,pattern);
            pattern.pop_back();
        }
    }
};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        List<Character> temp = new ArrayList();
        Helper( n , 0, 0, ans,temp);
        return ans;
    }
    public void Helper(int n, int leftP, int rightP, List<String> ans, List<Character>temp){
        if(leftP==rightP && n==leftP) {
            StringBuilder sb = new StringBuilder("");
            for(char c : temp) sb.append(c);
            ans.add(sb.toString());
            return;
        }
        if(leftP<n)
        {
            temp.add('(');
            Helper(n, leftP+1,rightP,ans,temp);
            temp.remove(temp.size()-1);
        }
        if(rightP<leftP)
        {
            temp.add(')');
            Helper(n, leftP,rightP+1,ans,temp);
            temp.remove(temp.size()-1);
        }
        return;
    }
}